python/ray/data/dataset.py

```diff
@@ -397,24 +397,116 @@ def split(self,
         if n <= 0:
             raise ValueError(f"The number of splits {n} is not positive.")
 
-        if n > self.num_blocks() and equal:
-            raise NotImplementedError(
-                f"The number of splits {n} > the number of dataset blocks "
-                f"{self.num_blocks()}, yet an equal split was requested.")
-
         if locality_hints and len(locality_hints) != n:
             raise ValueError(
                 f"The length of locality_hints {len(locality_hints)} "
                 "doesn't equal the number of splits {n}.")
 
-        # TODO(ekl) we could do better than truncation here. This could be a
-        # problem if block sizes are very skewed.
-        def equalize(splits: List[Dataset[T]]) -> List[Dataset[T]]:
+        def equalize(splits: List[Dataset[T]], n: int) -> List[Dataset[T]]:
```

Comment:
This would be easier to read if broken down into smaller sub-routines, so the main body of equalize() is relatively high level (10-20LOC).

Reply:
right, i'm more thinking something like this
```
        def equalize(splits: List[Dataset[T]]) -> List[Dataset[T]]:
            if not equal or len(splits) == 0:
                return splits

            target_size = sum([s.count() for s in splits]) / len(splits) 
            left_over = None

            # first round: trim the large split
            for i, s in enumerate(splits):
                if s.count() <= target_size:
                    continue                
                s, remainder = s.split_at_indices([target_size])
                splits[i] = s
                left_over = left_over.union(remainder) if left_over else remainder

            if not left_over or left_over.count() == 0:
                return

            # second round: fill the small split
            for i, s in enumerate(splits):
                if s.count() == target_size:
                    continue                
                required = target_size - s.count()
                addition, left_over = left_over.split_at_indices([required])
                splits[i] = s.union(addition)
            
            return splits
```

Reply:
@scv119 A few reasons that I didn't go that route:
1. `s.count()` can be expensive (a fan-out of tasks to compute the number of rows in each block), so we should definitely cache these values instead of recomputing them.
2. We want to batch splits if we can since these are pretty expensive (involving tasks and most likely data transfer). E.g. for the case of `[1, 1, 1, 2, 20]`, the multi-spit and union buffer that I added would result in 1 union call and 1 split_at_indices call; what you have above would result in ~7 union calls and ~7 split_at_indices calls, and the choice of _how_ the datasets are split and merged is very suboptimal.
3. The number of input splits won't equal the number of output splits if `num_splits > len(ds._blocks)` (which I believe we will want to support), so we can't replace the original splits array.

Reply:
for 
1) does it make sense to optimize it at dataset/block (i.e. cache the count() result ) level? 
2) makes sense; 
3) not sure we need that.

Reply:
1. If we consider that in scope for this, then I can add that! I wanted to limit the scope of this PR to get it across the line, but that shouldn't be difficult to add.
3. I think that not allowing `num_splits > len(ds._blocks)` will result in pretty poor UX, since users will definitely try to do it and we'll be needlessly forcing them to repartition their dataset or reduce the number of consumers over a limitation on the number of blocks, which is mostly a Datasets implementation detail.

Reply:
let's defer 1) for later PR :)